### ISSUE TITLE
patricia.c and patricia.h are misleading titles

### ISSUE THREAD

**Issue by Sebbyastian** [atheme/libmowgli-2#35]
In order to find out what PATRICIA is, we need to trace it back to the origin of the term, as defined by D. R. Morrison in "Practical Algorithm to Retrieve Information Coded in Alphanumeric" which was published in the Journal of ACM in 1968. You should be able to find this document by Google searching "p514-morrison.pdf", though I'm not sure of the legality. I purchased my copy, and I encourage you to, also. It isn't the only document defining PATRICIA this way. I also recommend reading:

 - D.E. Knuth. The Art of Computer Programming Vol. 3, all of section 6.3.
 - R. Sedgewick. Algorithms in C. Addison-Wesley, edn 1, pp253, 1990.

I'll focus on the original paper, however, which outlines some very specific details about the data structure and algorithms denoted by PATRICIA, such as flowcharts, for example, though I'll stick to some of the textual descriptions:

> Early in this evolution it was decided that the alphabet should be restricted to a binary one.

PATRICIA takes the structure of a binary tree. There are requirements that follow on from this, but the most significant without a doubt is this:

>  These facts are very useful in the allocation of storage for the index. They imply that the total storage required is completely determined by the number of ends, and all of the storage required will actually be used.

Every node should be a leaf.

In conclusion, I think the name of these algorithms, the collections that they create and the files they're expressed within should be renamed so as to avoid future confusion.

**Comment by kaniini** [comment]
Hello,

Thank you for your feedback.

Mowgli's PATRICIA implementation began as an implementation of the original PATRICIA algorithm (well, the version described by Knuth in ACP volume 3, section 6.3 as you cite above).  For many years, this was the case. 

Then @jillest realized that it was possible to use nibbles instead to improve performance and tree depth while retaining API compatibility.  So it is true that the tree implemented in mowgli_patricia is not a true PATRICIA implementation anymore.  And it is still a type of radix tree, so while confusing, it's roots are clearly in the original PATRICIA algorithm.

With that said, Mowgli's purpose is not to be a teaching tool, it's to deliver a high quality, performant runtime for the Atheme stack.  Mowgli's patricia code is meant to provide a string-backed key to pointer mapping that is highly performant, not to teach people how to write a binary tree implementation (and as you mention, we use nibbles now).

So while I think the documentation could be improved to clarify that this is no longer strictly a PATRICIA tree implementation, the API itself should remain as it is for the sake of compatibility with its consumers.

As such, closing this bug.  If you disagree with my logic you can reply, but please do not reopen.
